File: haskell-stm/src/Transfer3.hs
module Transfer3
  ( main3
  )
where

import           Control.Monad                  ( when )
import           Control.Concurrent.STM
import           Control.Concurrent             ( forkIO
                                                , threadDelay
                                                )

{-
 -3rd example on 'choice' using `orElse`
 -(limitedWithdraw acc1 acc2 amt) withdraws amt from acc1,
 -if acc1 has enough money, otherwise from acc2.
 -If neither has enough, it retries.
 -}

type Account = TVar Int
type Name = String

withdraw :: Account -> Int -> STM ()
withdraw acc amount = do
  bal <- readTVar acc
  check (amount <= 0 || amount <= bal)
  writeTVar acc (bal - amount)

limitedWithdraw :: Account -> Account -> Int -> STM ()
limitedWithdraw acc1 acc2 amt =
  withdraw acc2 amt `orElse` withdraw acc2 amt

delayDeposit :: Name -> Account -> Int -> IO ()
delayDeposit name acc amount = do
  threadDelay 3000000
  putStrLn ("Depositing $" ++ show amount ++ " into " ++ name)
  atomically
    (do
      bal <- readTVar acc
      writeTVar acc (bal + amount)
    )

showAcc :: Name -> Account -> IO ()
showAcc name acc = do
  bal <- readTVarIO acc
  putStrLn (name ++ ": $" ++ show bal)

main3 = do
  acc1 <- atomically (newTVar 100)
  acc2 <- atomically (newTVar 100)
  showAcc "Left pocket"  acc1
  showAcc "Right pocket" acc2
  forkIO (delayDeposit "Right pocket" acc2 1)
  putStrLn "Withdrawing $101 from either pocket..."
  atomically (limitedWithdraw acc1 acc2 101)
  putStrLn "Successful withdrawal!"
  showAcc "Left pocket"  acc1
  showAcc "Right pocket" acc2

File: scala-stm/src/main/scala/io/github/gvolpe/transfer3.scala
package io.github.gvolpe

import scalaz.zio._
import scalaz.zio.clock.Clock
import scalaz.zio.duration._
import scalaz.zio.stm._
import transfer._

object transfer3 {

  def withdraw(account: Account, amount: Int): STM[Nothing, Unit] =
    account.ref.get.flatMap { balance =>
      STM.check(amount <= 0 || amount <= balance) *>
        account.ref.set(balance - amount)
    }

  def limitedWithdraw(acc1: Account, acc2: Account, amount: Int): STM[Nothing, Unit] =
    withdraw(acc1, amount).orElse(withdraw(acc2, amount))

  def delayDeposit(name: Name, account: Account, amount: Int): ZIO[Clock, Nothing, Unit] =
    for {
      _ <- ZIO.sleep(3.seconds)
      _ <- putStrLn("Depositing $" + amount + s" into ${name.value}")
      _ <- STM.atomically(account.ref.update(_ + amount))
    } yield ()

  def showAccount(name: Name, account: Account): UIO[Unit] =
    account.ref.get.commit.flatMap { balance =>
      putStrLn(name.value + ": $" + balance)
    }

  val leftPocket = Name("Left pocket")
  val rightPocket = Name("Right pocket")

  def main: ZIO[Clock, Nothing, Unit] =
    for {
      acc1 <- STM.atomically(TRef.make(100)).map(Account)
      acc2 <- STM.atomically(TRef.make(100)).map(Account)
      _    <- showAccount(leftPocket, acc1)
      _    <- showAccount(rightPocket, acc2)
      _    <- delayDeposit(rightPocket, acc2, 1).fork.void
      _    <- putStrLn("Withdrawing $101 from either pocket...")
      _    <- STM.atomically(limitedWithdraw(acc1, acc2, 101))
      _    <- putStrLn("Successful withdrawal!")
      _    <- showAccount(leftPocket, acc1)
      _    <- showAccount(rightPocket, acc2)
    } yield ()

}

